server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -284,68 +281,23 @@ private static String getExceptionMessage(Throwable e) {
+            if (item.seqNo() != SequenceNumbers.SKIP_ON_REPLICA) {
+                return false;
             }
-            SourceToParse sourceToParse = new SourceToParse(
-                request.index(),
-                item.id(),
-                item.source(),
-                XContentType.JSON
-            );
-            IndexResult result = indexShard.applyIndexOperationOnReplica(
-                item.seqNo(),
-                item.primaryTerm(),
-                item.version(),
-                Translog.UNSET_AUTO_GENERATED_TIMESTAMP,
-                false,
-                sourceToParse
-            );
-            if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
-                // Even though the primary waits on all nodes to ack the mapping changes to the master
-                // (see MappingUpdatedAction.updateMappingOnMaster) we still need to protect against missing mappings
-                // and wait for them. The reason is concurrent requests. Request r1 which has new field f triggers a
-                // mapping update. Assume that that update is first applied on the primary, and only later on the replica
-                // (it’s happening concurrently). Request r2, which now arrives on the primary and which also has the new
-                // field f might see the updated mapping (on the primary), and will therefore proceed to be replicated
-                // to the replica. When it arrives on the replica, there’s no guarantee that the replica has already
-                // applied the new mapping, so there is no other option than to wait.
-                throw new TransportReplicationAction.RetryOnReplicaException(indexShard.shardId(),
-                    "Mappings are not available on the replica yet, triggered update: " + result.getRequiredMappingUpdate());
-            }
-            location = result.getTranslogLocation();
         }
-        return new WriteReplicaResult<>(location, null, indexShard);
+        return true;
     }
 
     @Override
     protected WriteReplicaResult<ShardUpsertRequest> processRequestItemsOnReplica(IndexShard indexShard, ShardUpsertRequest request) throws IOException {
         Reference[] insertColumns = request.insertColumns();
         if (insertColumns == null) {
-            return legacyReplicaOp(indexShard, request);
-        }
-        // If an item has a source, parse it instead of using Indexer
-        // The request may come from an older node in a mixed cluster.
-        // In that case the insertColumns/values won't always include undeterministic generated columns
-        // and we'd risk generating new diverging values on the replica
-        for (var item : request.items()) {
-            if (item.source() != null) {
-                return legacyReplicaOp(indexShard, request);
-            }
+            assert shouldIgnoreAllItems(request);
+            return new WriteReplicaResult<>(null, null, indexShard);
```

Comment:
I think it would be worth adding a comment that explains why this scenario can happen, and why it's okay to handle it like this.

Reply:
This took me a while to work out!  Updates on the primary get rewritten as inserts, so on the replica there will always be insert columns defined unless there are no items to index, or all the items failed on the primary.  Have added a comment to make this clear.